manpages/testdc.1
.TH TESTDC "7" "October 2012" "CompatibleOne" "Platform"
.SH NAME
testdc \- Test DeltaCloud
.SH SYNOPSIS
\fBtestdc\fR
.PP
.SH DESCRIPTION
.\" Add any additional description here
.PP
This will run the CompatibleOne \fBtestdc\fR service.
.SH AUTHOR
The CompatibleOne project.
.SH MANPAGE
Most of this documents comes from the README of ACCORDS, written by Iain James Marshall.
Written by Alexandre Lissy <alissy@mandriva.com> for the CompatibleOne project.
.SH "REPORTING BUGS"
Report CompatibleOne bugs to compatibleone@ow2.org
.br
CompatibleOne home page: <http://www.compatibleone.org>
.SH COPYRIGHT
Copyright \(co 2012 CompatibleOne.
License Apache 2.0.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
compatibleone-accords (0.20121218-1) unstable; urgency=low

  * New upstream code + more doc + sync all binary packages.

 -- Vincent Renardias <vincent@renardias.com>  Tue, 18 Dec 2012 14:43:00 +0200

compatibleone-accords (0.20121214-1) unstable; urgency=low

  * Several bug fixes regarding cosacs.

 -- Vincent Renardias <vincent@renardias.com>  Fri, 14 Dec 2012 16:58:00 +0200

compatibleone-accords (0.20121213-2) unstable; urgency=low

  * Add 'cosacs' package (client side COSACS to be run in the VM images).
  * Split 'compatibleone' in 'compatibleone-base' and 'compatibleone'.

 -- Vincent Renardias <vincent@renardias.com>  Thu, 13 Dec 2012 17:47:00 +0200

compatibleone-accords (0.20121213-1) unstable; urgency=low

  * Major renaming of packages to be fully Debian compliant and prepare the
    integration of the Cosacs client.
  * Force lib # in the dependency and bump the version to make sure we're
    not linking against a version older than today (lib ABI change).

 -- Vincent Renardias <vincent@renardias.com>  Tue, 12 Dec 2012 15:16:00 +0200

compatibleone-accords (0.20121212-1) unstable; urgency=low

  * Upstream update.
  * Major packaging update: each lib now has its own packages.

 -- Vincent Renardias <vincent@renardias.com>  Tue, 11 Dec 2012 15:47:00 +0200

compatibleone-accords (0.20121207-1) unstable; urgency=low

  * Upstream update.
  * Lots of packaging fixes (see gitorious log for the gory details)

 -- Vincent Renardias <vincent@renardias.com>  Fri, 07 Dec 2012 10:25:21 +0200

compatibleone-accords (0.20121129-1) unstable; urgency=low

  * Upstream update.

 -- Vincent Renardias <vincent@renardias.com>  Mon, 26 Nov 2012 11:43:34 +0200

compatibleone-accords (0.20121121-1) unstable; urgency=low

  * Uptream update 2012/11/21 codebase + python enabled.
  * compatibleone-example and compatibleone-testosocci are now gone.

 -- Vincent Renardias <vincent@renardias.com>  Wed, 21 Nov 2012 11:12:59 +0200

compatibleone-accords (0.20121120-1) unstable; urgency=low

  * Uptream update 2012/11/20 codebase + perm fixes.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Tue, 20 Nov 2012 15:24:44 +0200

compatibleone-accords (0.20121114-2) unstable; urgency=low

  * Uptream update 2012/11/19 codebase.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Mon, 19 Nov 2012 20:41:06 +0200

compatibleone-accords (0.20121114-1) unstable; urgency=low

  * Uptream update,
  * jpaasprocci -> paasprocci renaming.
  * fixes for java build on 64bit arch.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Wed, 14 Nov 2012 14:24:57 +0200

compatibleone-accords (0.20121022-1) unstable; urgency=low

  * Upstream change: new accords bash script.
  * All executables now have a manpage.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Mon, 22 Oct 2012 15:18:00 +0200

compatibleone-accords (0.20121015-1) unstable; urgency=low

  * Only recommend compatibleone-paprocci so everybody isn't forced to install the Java runtime.
  * psmisc dep so that we're sure killall is installed.
  * Change OpenSSL dependancy so it also builds on Debian 6.0.x
  * Change debhelper dep so the package builds on Ubuntu 9.04.
  * Commit work with Mauricio to build compatibleone-paprocci and run it with
    the appropriate autodetected java environment.
  * co-stop: can now be run from any directory.
  * Add package compatibleone-example.
  * Add compatibleone-nagios package.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Mon, 15 Oct 2012 11:08:00 +0200

compatibleone-accords (0.20121008-1) unstable; urgency=low

  * More fixes/updates (see gitorious log).

 -- Vincent Renardias <vrenardias@citypassenger.com>  Mon, 8 Oct 2012 14:38:16 +0200

compatibleone-accords (0.20121004-1) unstable; urgency=low

  * More fixes/updates, Initial init script for publisher, WIP.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Thu, 4 Oct 2012 11:55:16 +0200

compatibleone-accords (0.20120918-1) unstable; urgency=low

  * Various fixes/updates, WIP.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Tue, 18 Sep 2012 11:03:43 +0200

compatibleone-accords (0.20120901-1) unstable; urgency=low

  * Initial Debian packaging of CompatibleOne Accords plateform.

 -- Vincent Renardias <vincent@renardias.com>  Sat, 01 Sep 2012 20:33:43 +0200


testdc/src/testdc.c
#ifndef	_testdc_c
#define	_testdc_c

#include "standard.h"
#include "dcclient.h"

private	int	verbose=0;
private	int	debug=0;
private	char *	publisher=(char *) 0;
private	char *	operator="accords";
private	char *	zone="europe";
private	char *	tls=(char *) 0;

public	int	check_debug()		{	return( debug);		}
public	int	check_verbose()		{	return( verbose);	}
public	char *	default_zone()		{	return( zone );		}
public	char *	default_operator()	{	return( operator );	}
public	char *	default_tls()		{	return( tls );		}
public	int	failure(int status, char * message)
{
	if ( status )
		printf("**failure(%u,%s)**\n",status,message);
	return( status );
}

public	char *	default_publisher()	
{
	return( publisher );
}

private	int	dc_result( struct rest_response* rptr )
{
	rptr = liberate_rest_response( rptr );
	return( 0 );
}

private	int	dc_operation( char * host, char * user, char * password, char * command, char * subject, char * option, char * option2 )
{
	if (!( host ))
		return( 30 );
	else if (!( user ))
		return( 31 );
	else if (!( password ))
		return( 32 );
	else if (!( command ))
		return( 33 );
	else if (!( subject ))
		return( 34 );
	else if (!( strcmp( command, "list" ) ))
	{
		if (!( strcmp( subject, "features" ) ))
			return( dc_result( dc_list_features() ) );
		else if (!( strcmp( subject, "profiles" ) ))
			return( dc_result( dc_list_profiles() ) );
		else if (!( strcmp( subject, "images" ) ))
			return( dc_result( dc_list_images() ) );
		else if (!( strcmp( subject, "storage" ) ))
			return( dc_result( dc_list_storage() ) );
		else if (!( strcmp( subject, "servers" ) ))
			return( dc_result( dc_list_instances() ) );
		else if (!( strcmp( subject, "keys" ) ))
			return( dc_result( dc_list_keys() ) );
		else if (!( strcmp( subject, "addresses" ) ))
			return( dc_result( dc_list_addresses() ) );
		else if (!( strcmp( subject, "firewalls" ) ))
			return( dc_result( dc_list_firewalls() ) );
		else if (!( strcmp( subject, "buckets" ) ))
			return( dc_result( dc_list_buckets() ) );
		else if (!( strcmp( subject, "loadbalancers" ) ))
			return( dc_result( dc_list_loadbalancers() ) );
		else	return( failure( 35, subject ) );
	}

	else if (!( strcmp( command, "get" ) ))
	{
		if (!( option ))
			return( failure( 36, "option") );
		else if (!( strcmp( subject, "profile" ) ))
			return( dc_result( dc_get_profile(option) ) );
		else if (!( strcmp( subject, "image" ) ))
			return( dc_result( dc_get_image(option) ) );
		else if (!( strcmp( subject, "storage" ) ))
			return( dc_result( dc_get_storage(option) ) );
		else if (!( strcmp( subject, "server" ) ))
			return( dc_result( dc_get_instance(option) ) );
		else if (!( strcmp( subject, "key" ) ))
			return( dc_result( dc_get_key(option) ) );
		else if (!( strcmp( subject, "addresse" ) ))
			return( dc_result( dc_get_address(option) ) );
		else if (!( strcmp( subject, "firewall" ) ))
			return( dc_result( dc_get_firewall(option) ) );
		else if (!( strcmp( subject, "bucket" ) ))
			return( dc_result( dc_get_bucket(option) ) );
		else if (!( strcmp( subject, "blob" ) ))
			return( dc_result( dc_get_blob_info(option,option2) ) );
		else if (!( strcmp( subject, "loadbalancer" ) ))
			return( dc_result( dc_get_loadbalancer(option) ) );
		else	return( failure( 35, subject ) );
	}

	else if (!( strcmp( command, "create" ) ))
	{
		if (!( option ))
			return( failure( 36, "option") );
		else if (!( strcmp( subject, "image" ) ))
			return( dc_result( dc_create_image(option) ) );
		else if (!( strcmp( subject, "key" ) ))
			return( dc_result( dc_create_key(option) ) );
		else if (!( strcmp( subject, "server" ) ))
			return( dc_result( dc_create_instance(option) ) );
		else if (!( strcmp( subject, "addresse" ) ))
			return( dc_result( dc_create_address(option) ) );
		else if (!( strcmp( subject, "firewall" ) ))
			return( dc_result( dc_create_firewall(option) ) );
		else if (!( strcmp( subject, "loadbalancer" ) ))
			return( dc_result( dc_create_loadbalancer(option) ) );
		else	return( failure( 35, subject ) );
	}

	else if (!( strcmp( command, "delete" ) ))
	{
		if (!( option ))
			return( failure( 36, "option") );
		else if (!( strcmp( subject, "image" ) ))
			return( dc_result( dc_delete_image(option) ) );
		else if (!( strcmp( subject, "storage" ) ))
			return( dc_result( dc_delete_storage(option) ) );
		else if (!( strcmp( subject, "server" ) ))
			return( dc_result( dc_delete_instance(option) ) );
		else if (!( strcmp( subject, "key" ) ))
			return( dc_result( dc_delete_key(option) ) );
		else if (!( strcmp( subject, "addresse" ) ))
			return( dc_result( dc_delete_address(option) ) );
		else if (!( strcmp( subject, "firewall" ) ))
			return( dc_result( dc_delete_firewall(option) ) );
		else if (!( strcmp( subject, "rule" ) ))
			return( dc_result( dc_delete_rule(option,option2) ) );
		else if (!( strcmp( subject, "bucket" ) ))
			return( dc_result( dc_delete_bucket(option) ) );
		else if (!( strcmp( subject, "blob" ) ))
			return( dc_result( dc_delete_blob(option,option2) ) );
		else if (!( strcmp( subject, "loadbalancer" ) ))
			return( dc_result( dc_delete_loadbalancer(option) ) );
		else	return( failure( 35, subject ) );
	}

	else if (!( strcmp( command, "start" ) ))
	{
		if (!( strcmp( subject, "server" ) ))
			return( dc_result( dc_start_instance(option) ) );
	}

	else if (!( strcmp( command, "stop" ) ))
	{
		if (!( strcmp( subject, "server" ) ))
			return( dc_result( dc_stop_instance(option) ) );
	}

	else if (!( strcmp( command, "attach" ) ))
	{
		if (!( strcmp( subject, "storage" ) ))
			return( dc_result( dc_attach_storage(option,option2) ) );
	}

	else if (!( strcmp( command, "detach" ) ))
	{
		if (!( strcmp( subject, "storage" ) ))
			return( dc_result( dc_detach_storage(option) ) );
	}

	else if (!( strcmp( command, "register" ) ))
	{
		if (!( strcmp( subject, "loadbalancer" ) ))
			return( dc_result( dc_register_loadbalancer(option,option2) ) );
	}

	else if (!( strcmp( command, "unregister" ) ))
	{
		if (!( strcmp( subject, "loadbalancer" ) ))
			return( dc_result( dc_unregister_loadbalancer(option,option2) ) );
	}

	else if (!( strcmp( command, "associate" ) ))
	{
		if (!( strcmp( subject, "address" ) ))
			return( dc_result( dc_associate_address(option,option2) ) );
	}

	else if (!( strcmp( command, "disassociate" ) ))
	{
		if (!( strcmp( subject, "address" ) ))
			return( dc_result( dc_disassociate_address(option) ) );
	}

	else	return( failure( 37, command ) );

	return(0);
}

private int	dc_create_command( char * subject, int argi, int argc, char * argv[] )
{
	char *	filename;
	int	argn=0;
	char *	args[10];
	for ( argn=0; argn < 10; argn++ ) args[argn] = (char *) 0;
	argn=0;
	while ( argi < argc )
		args[argn++] = argv[argi++];

	if (!( strcmp( subject, "server" ) ))
	{
		if (!( filename = dc_create_instance_message( args[0], args[1], args[2], args[3], args[4] ) ))
			return( 46 );
		else 	return( dc_result( dc_create_instance( filename ) ) );
	}
	else if (!( strcmp( subject, "rule" ) ))
	{
		if (!( filename = dc_create_rule_message(  args[1], args[2], args[3], args[4] ) ))
			return( 46 );
		else 	return( dc_result( dc_create_rule( args[0], filename ) ) );
	}
	else if (!( strcmp( subject, "loadbalancer" ) ))
	{
		if (!( filename = dc_create_loadbalancer_message( args[0], args[1], args[2], args[3], args[4] ) ))
			return( 46 );
		else 	return( dc_result( dc_create_loadbalancer( filename ) ) );
	}
	else if (!( strcmp( subject, "key" ) ))
	{
		if (!( filename = dc_create_key_message( args[0] ) ))
			return( 46 );
		else 	return( dc_result( dc_create_key( filename ) ) );
	}
	else if (!( strcmp( subject, "image" ) ))
	{
		if (!( filename = dc_create_image_message( args[0], args[1], args[2] ) ))
			return( 46 );
		else 	return( dc_result( dc_create_image( filename ) ) );
	}
	else if (!( strcmp( subject, "storage" ) ))
	{
		if (!( filename = dc_create_storage_message( args[0], args[1], args[2] ) ))
			return( 46 );
		else 	return( dc_result( dc_create_storage( filename ) ) );
	}
	else if (!( strcmp( subject, "firewall" ) ))
	{
		if (!( filename = dc_create_firewall_message( args[0], args[1] ) ))
			return( 46 );
		else 	return( dc_result( dc_create_firewall( filename ) ) );
	}
	else if (!( strcmp( subject, "bucket" ) ))
	{
		if (!( filename = dc_create_bucket_message( args[0], args[1] ) ))
			return( 46 );
		else 	return( dc_result( dc_create_bucket( filename ) ) );
	}
	else if (!( strcmp( subject, "blob" ) ))
	{
		return( dc_result( dc_update_blob( args[0], args[1], args[2] ) ) );
	}
	else if (!( strcmp( subject, "address" ) ))
	{
		return( dc_result( dc_create_address() ) );
	}
	else	return( 30 );

}

private int	dc_test_main( int argc, char * argv[] )
{
	char *	host=(char *) 0;
	char *	user=(char *) 0;
	char * 	tenant=(char *) 0;
	char *	agent="deltacloud-client/1.0";
	char *	password=(char *) 0;
	char *	command=(char *) 0;
	char *	subject=(char *) 0;
	int	argi=1;
	char * aptr;
	while ( argi < argc )
	{
		if (!( aptr = argv[argi++] ))
			break;
		else if ( *aptr != '-' )
		{
			dc_api_configuration( host, user, password, tenant, agent,tls );
			if (!( command ))
				command = aptr;
			else if (!( subject ))
			{
				subject = aptr;
				if (!( strcmp( command, "list" ) ))
					return( dc_operation( host, user, password, command, subject, (char *) 0,(char *) 0 ) );
				else if (!( strcmp( command, "create" ) ))
					return( dc_create_command( subject, argi, argc, argv ) );
			}
			else	return( dc_operation( host, user, password, command, subject, aptr, argv[argi]) );
		}
		else if ( *(aptr+1) = '-' )
		{
			aptr+=2;
			if (!( strcmp(aptr,"host" )))
				host = argv[argi++];
			else if (!( strcmp(aptr,"user" )))
				user = argv[argi++];
			else if (!( strcmp(aptr,"password" )))
				password = argv[argi++];
			else if (!( strcmp(aptr,"tenant" )))
				tenant = argv[argi++];
			else if (!( strcmp(aptr,"tls" )))
				tls = argv[argi++];
			else if (!( strcmp( aptr, "verbose" ) ))
				verbose =1;
			else if (!( strcmp( aptr, "debug" ) ))
				debug =1;
			else
			{
				printf("\nincorrect option: %s\n",aptr);
				return(30);
			}
		}
		else
		{
			printf("\nincorrect option: %s\n",aptr);
			return(30);
		}
	}
	return( 0 );
}

private int dc_test_banner()
{
	printf("\n   CompatibleOne Delta Cloud Test Client : Version 1.0a.0.01 ");
	printf("\n   Beta Version 14/05/2012 ");
	printf("\n   Copyright (c) 2012 Iain James Marshall\n");
	printf("\n   Options:\n");
	printf("\n   --host     {hostname}       specify delta cloud host ");
	printf("\n   --user     {username}       specify delta cloud user ");
	printf("\n   --password {password}       specify user password    ");
	printf("\n   --tls      {filename}       specify traansport layer security ");
	printf("\n   --verbose                   activate information messages ");
	printf("\n   --debug                     activate debug messages \n");
	printf("\n   Methods:\n");
	printf("\n   list   [ features | profiles | images | servers | keys | addresses | firewalls | loadbalancers ]");
	printf("\n   list   [ storage | buckets ] ");
	printf("\n   get    [ profile | storage | image | server | key | addresse | firewall | loadbalancer ] {id} ");
	printf("\n   delete [ storage | image | server | key | addresse | firewall | rule | loadbalancer ] {id} ");
	printf("\n   create server name image profile firewall realm ");
	printf("\n   create loadbalancer name realm protocol port1 port2 ");
	printf("\n   create image instance name description ");
	printf("\n   create firewall name description ");
	printf("\n   create storage name capacity realm ");
	printf("\n   create rule firewall proto from to range ");
	printf("\n   create key name ");
	printf("\n   create address ");
	printf("\n   get    bucket id ");
	printf("\n   get    blob   bucketid blobid ");
	printf("\n   create bucket name location ");
	printf("\n   create blob   bucketid blobid filename ");
	printf("\n   delete bucket id "); 
	printf("\n   delete blob   bucketid blobid \n");
	printf("\n   Actions:\n");
	printf("\n   start        server id ");
	printf("\n   stop         server id ");
	printf("\n   attach       storage id server ");
	printf("\n   detach       storage id ");
	printf("\n   register     loadbalancer id server ");
	printf("\n   unregister   loadbalancer id server ");
	printf("\n   associate    address id server ");
	printf("\n   disassociate address id ");
	printf("\n\n");
	return( 0 );
}

public int	main( int argc, char * argv[] )
{
	int	status;
	if ( argc == 1 )
		return( dc_test_banner() );
	else if (!( status = dc_test_main(argc, argv) ))
		return(status);
	else
	{
		printf("**failure(%u,%s)**\n",status,argv[0]);
		return(0);
	}
}

#endif 	/* _testdc_c */
	/* -------- */

